File: ios/WealthBoard/Data/ExchangeZones.swift
import Foundation

/// The IANA time zone a listing actually trades in, derived from its ticker.
/// Ported from `data/ExchangeZones.kt`.
///
/// A quote endpoint reports `meta.exchangeTimezoneName` and that is always
/// preferred — but it is not always there. Batched price calls carry no zone at
/// all. Without a fallback, those rows print their last-trade time on the
/// *device's* clock, which is how the FTSE's 4:30 p.m. London close ends up
/// rendered as "11:30 a.m." for a reader in Toronto: a correct instant, labelled
/// in a way that reads as though London trades through the morning.
///
/// The suffix groups mirror `TickerFlag` exactly, so a ticker that gets a flag
/// also gets a zone.
enum ExchangeZones {

    /// Default for anything with no suffix — unsuffixed symbols are US.
    private static let us = "America/New_York"

    /// Zones for the index symbols the app lists, which have no suffix to read.
    /// Suffixed indices (FTSEMIB.MI, 000001.SS, ^SET.BK …) resolve through the
    /// suffix table below and are deliberately absent here.
    private static let indexZones: [String: String] = [
        "^DJI": us,
        "^GSPC": us,
        "^IXIC": us,
        "^RUT": us,
        "^VIX": us,
        "^NYA": us,
        "^GSPTSE": "America/Toronto",
        "^MXX": "America/Mexico_City",
        "^BVSP": "America/Sao_Paulo",
        "^FTSE": "Europe/London",
        "^GDAXI": "Europe/Berlin",
        "^STOXX50E": "Europe/Berlin",
        "^FCHI": "Europe/Paris",
        "^AEX": "Europe/Amsterdam",
        "^IBEX": "Europe/Madrid",
        "^SSMI": "Europe/Zurich",
        "^OMX": "Europe/Stockholm",
        "^OSEAX": "Europe/Oslo",
        "^OMXC25": "Europe/Copenhagen",
        "^N225": "Asia/Tokyo",
        "^HSI": "Asia/Hong_Kong",
        "^KS11": "Asia/Seoul",
        "^TWII": "Asia/Taipei",
        "^STI": "Asia/Singapore",
        "^KLSE": "Asia/Kuala_Lumpur",
        "^JKSE": "Asia/Jakarta",
        "^NSEI": "Asia/Kolkata",
        "^BSESN": "Asia/Kolkata",
        "^AXJO": "Australia/Sydney",
        "^AORD": "Australia/Sydney",
        "^NZ50": "Pacific/Auckland"
    ]

    private static let suffixZones: [String: String] = [
        ".TO": "America/Toronto", ".V": "America/Toronto", ".NE": "America/Toronto",
        ".CN": "America/Toronto", ".TSX": "America/Toronto",
        ".MX": "America/Mexico_City",
        ".SA": "America/Sao_Paulo",
        ".L": "Europe/London",
        ".IR": "Europe/Dublin",
        ".DE": "Europe/Berlin", ".F": "Europe/Berlin", ".BE": "Europe/Berlin",
        ".SG": "Europe/Berlin", ".HM": "Europe/Berlin", ".DU": "Europe/Berlin",
        ".MU": "Europe/Berlin", ".HA": "Europe/Berlin",
        ".PA": "Europe/Paris",
        ".AS": "Europe/Amsterdam",
        ".BR": "Europe/Brussels",
        ".LS": "Europe/Lisbon",
        ".MC": "Europe/Madrid",
        ".MI": "Europe/Rome",
        ".SW": "Europe/Zurich", ".VX": "Europe/Zurich",
        ".VI": "Europe/Vienna",
        ".ST": "Europe/Stockholm",
        ".OL": "Europe/Oslo",
        ".CO": "Europe/Copenhagen",
        ".HE": "Europe/Helsinki",
        ".IC": "Atlantic/Reykjavik",
        ".WA": "Europe/Warsaw",
        ".IS": "Europe/Istanbul",
        ".TA": "Asia/Jerusalem",
        ".JO": "Africa/Johannesburg",
        ".T": "Asia/Tokyo", ".JP": "Asia/Tokyo",
        ".HK": "Asia/Hong_Kong",
        ".SS": "Asia/Shanghai", ".SZ": "Asia/Shanghai",
        ".KS": "Asia/Seoul", ".KQ": "Asia/Seoul",
        ".TW": "Asia/Taipei", ".TWO": "Asia/Taipei",
        ".SI": "Asia/Singapore",
        ".KL": "Asia/Kuala_Lumpur",
        ".JK": "Asia/Jakarta",
        ".BK": "Asia/Bangkok",
        ".PS": "Asia/Manila",
        ".NS": "Asia/Kolkata", ".BO": "Asia/Kolkata",
        ".AX": "Australia/Sydney",
        ".NZ": "Pacific/Auckland"
    ]

    /// Zone for `ticker`, or nil when the symbol says nothing about where it
    /// trades (crypto, FX, an unrecognised suffix).
    ///
    /// Crypto and FX deliberately return nil rather than a venue zone: they
    /// trade continuously with no home exchange, so the viewer's own clock is
    /// the honest label for them.
    static func forTicker(_ ticker: String?) -> String? {
        guard let raw = ticker?.trimmingCharacters(in: .whitespaces).uppercased(), !raw.isEmpty else {
            return nil
        }
        if raw.contains("-USD") || raw.contains("-CAD") || raw.contains("-EUR")
            || raw.hasSuffix("=X") || raw.contains("USD=") {
            return nil
        }

        // Index symbols carry no exchange suffix to read — "^GSPTSE" is as
        // opaque as "^GSPC" — so they are listed outright. Without this they
        // all fall through to the unsuffixed default and every foreign index on
        // the Markets tab is labelled in New York's hours.
        if let indexZone = indexZones[raw] { return indexZone }

        let t = raw.hasPrefix("^") ? String(raw.dropFirst()) : raw
        guard let dotIndex = t.lastIndex(of: ".") else {
            return us   // AAPL, ES=F → US venues
        }
        let suffix = String(t[dotIndex...])
        return suffixZones[suffix]
    }

    /// The zone to render a quote's timestamp in: what the provider reported,
    /// then what the ticker implies, then nil for "use the viewer's clock".
    ///
    /// A provider zone is validated against the system's own database rather
    /// than trusted — an id this device doesn't know would silently resolve to
    /// GMT and print an hour that is simply wrong.
    static func resolve(providerZone: String?, ticker: String?) -> String? {
        if let reported = providerZone?.trimmingCharacters(in: .whitespaces),
           !reported.isEmpty,
           TimeZone(identifier: reported) != nil {
            return reported
        }
        return forTicker(ticker)
    }
}

File: android/app/src/main/java/ca/tristan/portfolio/data/ExchangeZones.kt
package ca.tristan.portfolio.data

import java.util.TimeZone

/**
 * The IANA time zone a listing actually trades in, derived from its ticker.
 *
 * Yahoo's chart endpoint reports `meta.exchangeTimezoneName` and that is
 * always preferred — but it is not always there. The Dashboard's price refresh
 * runs through Finnhub for US symbols and through a batched `spark` call for
 * everything else, and neither payload carries a zone at all. Without a
 * fallback, those rows printed their last-trade time on the *device's* clock,
 * which is how the FTSE's 4:30 p.m. London close ended up rendered as
 * "11:30 a.m." for a reader in Toronto: a correct instant, labelled in a way
 * that reads as though London trades through the morning.
 *
 * The suffix groups mirror [ca.tristan.portfolio.ui.format.TickerFlag] exactly,
 * so a ticker that gets a flag also gets a zone.
 */
object ExchangeZones {

    /** Default for anything with no suffix — Yahoo's unsuffixed symbols are US. */
    private const val US = "America/New_York"


    /**
     * Zones for the index symbols the app lists, which have no suffix to read.
     * Suffixed indices (FTSEMIB.MI, 000001.SS, ^SET.BK …) resolve through the
     * suffix table below and are deliberately absent here.
     */
    private val INDEX_ZONES: Map<String, String> = mapOf(
        "^DJI"      to US,
        "^GSPC"     to US,
        "^IXIC"     to US,
        "^RUT"      to US,
        "^VIX"      to US,
        "^NYA"      to US,
        "^GSPTSE"   to "America/Toronto",
        "^MXX"      to "America/Mexico_City",
        "^BVSP"     to "America/Sao_Paulo",
        "^FTSE"     to "Europe/London",
        "^GDAXI"    to "Europe/Berlin",
        "^STOXX50E" to "Europe/Berlin",
        "^FCHI"     to "Europe/Paris",
        "^AEX"      to "Europe/Amsterdam",
        "^IBEX"     to "Europe/Madrid",
        "^SSMI"     to "Europe/Zurich",
        "^OMX"      to "Europe/Stockholm",
        "^OSEAX"    to "Europe/Oslo",
        "^OMXC25"   to "Europe/Copenhagen",
        "^N225"     to "Asia/Tokyo",
        "^HSI"      to "Asia/Hong_Kong",
        "^KS11"     to "Asia/Seoul",
        "^TWII"     to "Asia/Taipei",
        "^STI"      to "Asia/Singapore",
        "^KLSE"     to "Asia/Kuala_Lumpur",
        "^JKSE"     to "Asia/Jakarta",
        "^NSEI"     to "Asia/Kolkata",
        "^BSESN"    to "Asia/Kolkata",
        "^AXJO"     to "Australia/Sydney",
        "^AORD"     to "Australia/Sydney",
        "^NZ50"     to "Pacific/Auckland"
    )

    /**
     * Zone for [ticker], or null when the symbol says nothing about where it
     * trades (crypto, FX, an unrecognised suffix).
     *
     * Crypto and FX deliberately return null rather than a venue zone: they
     * trade continuously with no home exchange, so the viewer's own clock is
     * the honest label for them.
     */
    fun forTicker(ticker: String?): String? {
        val raw = ticker?.trim()?.uppercase() ?: return null
        if (raw.isEmpty()) return null
        if (raw.contains("-USD") || raw.contains("-CAD") || raw.contains("-EUR") ||
            raw.endsWith("=X") || raw.contains("USD=")
        ) return null

        // Index symbols carry no exchange suffix to read — "^GSPTSE" is as
        // opaque as "^GSPC" — so they are listed outright. Without this they
        // all fell through to the unsuffixed default and every foreign index
        // on the Markets tab was labelled in New York's hours.
        INDEX_ZONES[raw]?.let { return it }

        val t = raw.removePrefix("^")
        val dot = t.lastIndexOf('.')
        if (dot < 0) return US                       // AAPL, ES=F → US venues
        return when (t.substring(dot)) {
            ".TO", ".V", ".NE", ".CN", ".TSX" -> "America/Toronto"
            ".MX"                             -> "America/Mexico_City"
            ".SA"                             -> "America/Sao_Paulo"
            ".L"                              -> "Europe/London"
            ".IR"                             -> "Europe/Dublin"
            ".DE", ".F", ".BE", ".SG", ".HM",
            ".DU", ".MU", ".HA"               -> "Europe/Berlin"
            ".PA"                             -> "Europe/Paris"
            ".AS"                             -> "Europe/Amsterdam"
            ".BR"                             -> "Europe/Brussels"
            ".LS"                             -> "Europe/Lisbon"
            ".MC"                             -> "Europe/Madrid"
            ".MI"                             -> "Europe/Rome"
            ".SW", ".VX"                      -> "Europe/Zurich"
            ".VI"                             -> "Europe/Vienna"
            ".ST"                             -> "Europe/Stockholm"
            ".OL"                             -> "Europe/Oslo"
            ".CO"                             -> "Europe/Copenhagen"
            ".HE"                             -> "Europe/Helsinki"
            ".IC"                             -> "Atlantic/Reykjavik"
            ".WA"                             -> "Europe/Warsaw"
            ".IS"                             -> "Europe/Istanbul"
            ".TA"                             -> "Asia/Jerusalem"
            ".JO"                             -> "Africa/Johannesburg"
            ".T", ".JP"                       -> "Asia/Tokyo"
            ".HK"                             -> "Asia/Hong_Kong"
            ".SS", ".SZ"                      -> "Asia/Shanghai"
            ".KS", ".KQ"                      -> "Asia/Seoul"
            ".TW", ".TWO"                     -> "Asia/Taipei"
            ".SI"                             -> "Asia/Singapore"
            ".KL"                             -> "Asia/Kuala_Lumpur"
            ".JK"                             -> "Asia/Jakarta"
            ".BK"                             -> "Asia/Bangkok"
            ".PS"                             -> "Asia/Manila"
            ".NS", ".BO"                      -> "Asia/Kolkata"
            ".AX"                             -> "Australia/Sydney"
            ".NZ"                             -> "Pacific/Auckland"
            else                              -> null
        }
    }

    /**
     * The zone to render a quote's timestamp in: what the provider reported,
     * then what the ticker implies, then null for "use the viewer's clock".
     *
     * A provider zone is validated against the JVM's own database rather than
     * trusted — an id this device doesn't know would silently resolve to GMT
     * and print an hour that is simply wrong.
     */
    fun resolve(providerZone: String?, ticker: String?): String? {
        val reported = providerZone?.trim()?.takeIf { it.isNotEmpty() }
        if (reported != null && TimeZone.getAvailableIDs().contains(reported)) return reported
        return forTicker(ticker)
    }
}
